File: algorithms/js/graph/courseSchedule.js
// Restate: Find if a student can take all courses.
// Pattern: DFS graph traversal. Use a set to keep track of nodes visited in current path.
// Use another set if I want to optimize and avoid running duplicate traversal.
// Invariant: Keep track of nodes visited in current recursive path. Clean it up when done with traversing
// the starting node and it's neighbors.
// Action: Loop through the courses and map out in a dictionary what courses have what prerequisite.
// Loop through the dictionary and traverse through the node and its neighbors while keeping track of visiting
// and visited nodes. When a node has been visited, we can move on to the next prerequisite to traverse as we will be
// revisiting a path that we've already explored. If we are done with traversing the current prerequisite
// and its neighbors, we clear out the set as we are done with the current recursive call stack.
// Why: This works because we keep track of visited nodes and update them when needed. If we see a node
// that has already been visited, we've found a cycle in the graph so a student can't take all courses.
// If we dont, then we know we're good
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
const canFinish = (numCourses, prereqs) => {
    const visited = new Set();
    const visiting = new Set();
    const courseMap = new Array(numCourses).fill(0).map(() => []);

    for (let [course, prereq] of prereqs) {
        courseMap[course].push(prereq);
    }

    const dfs = (course) => {
        if (visited.has(course)) return true;
        if (visiting.has(course)) return false;

        visiting.add(course);

        for (let prereq of courseMap[course]) {
            if (!dfs(prereq)) return false;
        }

        visiting.delete(course);
        visited.add(course);

        return true;
    }

    for (let i = 0; i < numCourses; i++) {
        if (!dfs(i)) return false;
    }

    return true;
}

// prereq -> course
const isValidCourses = (numCourses, prereqs) => {
    const courseMap = new Map();

    for (let i = 0; i < numCourses; i++) {
        courseMap.set(i, []);
    }

    for ([course, prereq] of prereqs) {
        courseMap.get(course).push(prereq);
    }

    const state = new Array(numCourses).fill(0);

    const dfs = (course) => {
        if (state[course] === 1) return false;
        if (state[course] === 2) return true;

        state[course] = 1;

        for (let prereq of courseMap.get(course)) {
            if (!dfs(prereq)) return false;
        }

        state[course] = 2;

        return true;
    }

    for (let i = 0; i < numCourses; i++) {
        if (!dfs(i)) return false;
    }

    return true;
}

// course -> prereq
const isValidCourses2 = (numCourses, prereqs) => {
    const courseMap = new Map();

    for (let i = 0; i < numCourses; i++) {
        courseMap.set(i, []);
    }

    for (let [course, prereq] of prereqs) {
        courseMap.get(course).push(prereq);
    }

    const visiting = new Set();
    const visited = new Set();

    const dfs = (course) => {
        if (visiting.has(course)) return false;
        if (visited.has(course)) return true;

        visiting.add(course);

        for (let prereq of courseMap.get(course)) {
            if (!dfs(prereq)) return false;
        }

        visiting.delete(course);
        visited.add(course);

        return true;
    }

    for (let i = 0; i < numCourses; i++) {
        if (!dfs(i)) return false;
    }

    return true;
}


// Kahn's
const canFinish = (numCourses, prerequisites) => {
    const graph = Array.from({ length: numCourses }, () => []);
    const inDegree = new Array(numCourses).fill(0);

    for (let [course, prereq] of prerequisites) {
        graph[prereq].push(course);
        inDegree[course]++;
    }

    const queue = [];

    for (let i = 0; i < numCourses; i++) {
        if (inDegree[i] === 0) queue.push(i);
    }

    let completed = 0;

    while (queue.length) {
        const course = queue.shift();
        completed++;

        for (let next of graph[course]) {
            inDegree[next]--;
            if (inDegree[next] === 0) queue.push(next);
        }
    }

    return completed === numCourses;
};

const canFinish2 = (numCourses, prerequisites) => {
    const graph = new Array(numCourses).fill(0).map(() => []);
    const remainingPrereqs = new Array(numCourses).fill(0);

    for (let [course, prereq] of prerequisites) {
        // This is because in DFS, we go to the end of the last prerequisite. Ex: course -> prereq -> prereq...
        // In BFS, we start with prerequisite that has no remaining dependencies
        // Then tries to see which course unlocks. so prereq -> course
        graph[prereq].push(course);
        remainingPrereqs[course]++;
    }

    const queue = [];

    for (let course = 0; course < numCourses; course++) {
        if (remainingPrereqs[course] === 0) queue.push(course);
    }

    let completedCourses = 0;

    while (queue.length) {
        const course = queue.shift();
        completedCourses++;

        for (let unlockedCourse of graph[course]) {
            remainingPrereqs[unlockedCourse]--;
            if (remainingPrereqs[unlockedCourse] === 0) queue.push(unlockedCourse);
        }
    }

    return completedCourses === numCourses;
}

// [0 = [], 1 = [], 2 = []]

// [0,1] [1, 2] [2,3]
// {0: [1]
//  1: [2]
//. 2: [3]}

const canFinishDFS = (numCourses, prerequisites) => {
    const graph = new Array(numCourses).fill(0).map(() => []);
    const state = new Array(numCourses).fill(0);

    for (let [course, prereq] of prerequisites) {
        graph[course].push(prereq);
    }

    const dfs = (course) => {
        if (state[course] === 1) return false;
        if (state[course] === 2) return true;

        state[course] = 1;

        for (let prereq of graph[course]) {
            if (!dfs(prereq)) return false;
        }

        state[course] = 2;

        return true;
    };

    for (let course = 0; course < numCourses; course++) {
        if (!dfs(course)) return false;
    }

    return true;
}

File: algorithms/py/graph/courseSchedule.py
# There are a total of numCourses courses you have to take, labeled from 0 to numCourses - 1. You are given an array prerequisites where prerequisites[i] = [ai, bi] indicates that you must take course bi first if you want to take course ai.

# For example, the pair [0, 1], indicates that to take course 0 you have to first take course 1.
# Return true if you can finish all courses. Otherwise, return false.

# Pattern:
    # Topological sort with graph traversal

# Approach:
    # I create a mapping of courses and prerequisites. I can go two ways:
    #     - course -> prerequisites we keep hashsets to see if there's any circular reference
    #     - prereq -> courses (Kahn's algorithm) where I will find a prereq course without any prereqs 
    #                          then move to courses to reduce their prereq count by 1, continue to see if we can exhaust all courses

# Time & Space complexity:
#     Time complexity of both approaches are O(n) where n is the size of the numCourses
#     Space complexity of both appraoches are O(n) as well

# Standard DFS
class SolutionDFS:
    def courseSchedule(self, numCourses: int, prereqs: list[list[int]]) -> bool:
        courseMap = {}

        for course, prereq in prereqs:
            courseMap[course] = courseMap.get(course, [])
            courseMap[course].append(prereq)

        visited = set()
        visiting = set()

        def dfs(course):
            if course in visiting:
                return False
            
            if course in visited:
                return True

            visiting.add(course)

            if course in courseMap:
                for neighbor in courseMap[course]:
                    if not dfs(neighbor):
                        return False

            visiting.remove(course)
            visited.add(course)
            return True

        for course in range(numCourses):
            if dfs(course) == False:
                return False
        
        return True
            
# Kahn's Algorithm (BFS)
class SolutionBFS:
    def courseSchedule(self, numCourses: int, prereqs: list[list[int]]) -> bool:
        prereqMap = {}
        prereqFreqBucket = [0] * numCourses

        for course, prereq in prereqs:
            prereqMap[prereq] = prereqMap.get(prereq, [])
            prereqMap[prereq].append(course)

        for prereq, courses in prereqMap.items():
            for course in courses:
                prereqFreqBucket[course] += 1

        queue = []
        queueIndex = 0

        for i in range(len(prereqFreqBucket)):
            if prereqFreqBucket[i] == 0:
                queue.append(i)

        completed = 0

        while queueIndex < len(queue):
            course = queue[queueIndex]
            queueIndex += 1

            courses = prereqMap.get(course, [])
            for course in courses:
                prereqFreqBucket[course] -= 1

                if prereqFreqBucket[course] == 0:
                    queue.append(course)
            
            completed += 1

        return completed == numCourses